drivers: video: ov7670: Implement missing video API functions, controls and update init regs.

iabdalkader opened the pull request:
Add the missing `stream_start` and `stream_stop` API functions (the driver doesn't work without them) and implement video controls for horizontal mirror (hmirror) and vertical flip. Additionally, I've changed PCLK to free-running as the STM32 DCMI doesn't seem to work without it.

I would also like to change the default VSYNC/HSYNC polarity to high (0x03) to make this driver consistent with others. This change would allow us to reuse the same overlay without modifying vsync-active. Since this driver wasn't functioning at all, I assume it is safe to make this change as it won't break anything else?

josuah commented on drivers/video/ov7670.c:
```diff
@@ -547,10 +550,73 @@ static int ov7670_init(const struct device *dev)
+		return ret;
+	}
+
+	reg = enable ? (reg | OV76770_MVFP_MIRROR) : (reg & ~OV76770_MVFP_MIRROR);
+	return i2c_reg_write_byte_dt(&config->bus, OV7670_MVFP, reg);
+}
```
It might be convenient to have a "read/modify/write" function provided by Zephyr `video_common.c` some point in time... In the meantime, the way you did is the best IMHO.

josuah commented on drivers/video/ov7670.c:
```diff
@@ -204,11 +207,11 @@ static const struct video_format_cap fmts[] = {
-	{OV7670_MVFP, 0x20}, /* MVFP: Mirror/VFlip,Normal image */
+	{OV7670_MVFP, 0x00}, /* MVFP: Mirror/VFlip,Normal image */
 
 	/* configure the output timing */
-	/* PCLK does not toggle during horizontal blank, one PCLK, one pixel */
-	{OV7670_COM10, 0x20}, /* COM10 */
+	/* Free running PCLK, default VSYNC, HSYNC and PCLK */
+	{OV7670_COM10, 0x00}, /* COM10 */
```
I am just curious about it: Was this not working without this? Is that related to the flipping/mirroring?

josuah replied:
https://github.com/zephyrproject-rtos/zephyr/commit/3ce3ed38ba462bce1a7a625211da775a43c61351

FRDM-MCXN947: I have that board with me! If I find some time to test it, that would make things easy.

josuah replied:
Thank you for commenting further on this.

Good idea about unifying the signals polarity. An issue about it sounds like a good first step.

josuah replied:
cc @danieldegrasse just for visibility as you did the original driver.

josuah reviewed (commented):
I do not see a problem with it being merged like it is, but curious about what you think should be done about start()/stop() when it is not supported by the driver yet, to make sure I did not miss a bug elsewhere in the stack.

Thank you for this addition!

iabdalkader commented on drivers/video/ov7670.c:
```diff
@@ -547,10 +550,73 @@ static int ov7670_init(const struct device *dev)
+	return 0;
+}
+
+static int ov7670_stream_stop(const struct device *dev)
+{
+	return 0;
+}
```
> Which code was not working in case start/stop is not supported?

This fails if the op is not set:

https://github.com/zephyrproject-rtos/zephyr/blob/e4164963f22af072de43a6c56cbf0ad9f323ff1b/include/zephyr/drivers/video.h#L601

josuah reviewed (dismissed):
Thank you for clarifying the different points and making sure that the implications were intended.

WRT start()/stop() missing, it seems like the return status needs to be "it did nothing" rather than "it failed because it had nothing to be done". `return 0;` is one way to do that. If many sensors need that, then this could become part of the API in the future.

kartben commented on drivers/video/ov7670.c:
```diff
@@ -547,10 +550,73 @@ static int ov7670_init(const struct device *dev)
+	return 0;
+}
+
+static int ov7670_set_ctrl_vflip(const struct device *dev, bool enable)
+{
+	int ret;
+	uint8_t reg = OV7670_MVFP;
+	const struct ov7670_config *config = dev->config;
+
+	ret = i2c_write_dt(&config->bus, &reg, sizeof(reg));
+	if (ret < 0) {
+		return ret;
+	}
+	ret = i2c_read_dt(&config->bus, &reg, sizeof(reg));
+	if (ret < 0) {
+		return ret;
+	}
+
+	reg = enable ? (reg | OV76770_MVFP_VFLIP) : (reg & ~OV76770_MVFP_VFLIP);
+	return i2c_reg_write_byte_dt(&config->bus, OV7670_MVFP, reg);
+}
+
+static int ov7670_set_ctrl_hmirror(const struct device *dev, bool enable)
+{
+	int ret;
+	uint8_t reg = OV7670_MVFP;
+	const struct ov7670_config *config = dev->config;
+
+	ret = i2c_write_dt(&config->bus, &reg, sizeof(reg));
+	if (ret < 0) {
+		return ret;
+	}
+	ret = i2c_read_dt(&config->bus, &reg, sizeof(reg));
+	if (ret < 0) {
+		return ret;
+	}
+
+	reg = enable ? (reg | OV76770_MVFP_MIRROR) : (reg & ~OV76770_MVFP_MIRROR);
+	return i2c_reg_write_byte_dt(&config->bus, OV7670_MVFP, reg);
+}
```
That's a lot of duplicated code, can't there be a common helper function?

josuah replied:
Some sensors have a read-modify-write wrapper.
Similar review happened here, also in order to implement VFLIP/HFLIP:
- https://github.com/zephyrproject-rtos/zephyr/pull/81830#discussion_r1856545279

I hope this can be unified across all sensors in the long term. I will try something with IMX219 if time permits.

iabdalkader replied:
Updated 👍🏼

iabdalkader committed:
drivers: video: ov7670: Implement missing video API functions

Add the missing stream_start and stream_stop API functions
(the driver doesn't work without them) and implement video
controls for horizontal mirror (hmirror) and vertical flip.

Signed-off-by: Ibrahim Abdalkader <i.abdalkader@gmail.com>

iabdalkader committed:
drivers: video: ov7670: Use a free-running pixel clock.

The STM32 DCMI capture seems to stop when the pixel clock is
disabled during horizontal blank. This patch switches pixel
clock to free-running, which shouldn't have any effect on other
capture devices.

Signed-off-by: Ibrahim Abdalkader <i.abdalkader@gmail.com>

josuah commented:
Always good to get more things to work!

> but at least RGB won't have to be swapped in software.

Is this related to the image being swapped left/right per chance? This swaps the endianness for some sensors.

kartben commented:
> Can I fix this too while we're at it?

of course :) however it's usually not a bad idea to split things in multiple PRs to get incremental fixes/improvements approved (like, e.g. this PR here is now good to go, and you will be losing approvals if you push updates, and then folks might not be available to review again :) )

iabdalkader commented:
> Is this related to the image being swapped left/right per chance? This swaps the endianness for some sensors.

Yes this swaps the endianness. The current setting is set to UYVY. It seems to also swap RGB565 output.

![image](https://github.com/user-attachments/assets/b85af999-9697-4385-8f56-76cfa71a6449)

kartben commented on drivers/video/ov7670.c:
```diff
@@ -341,6 +344,27 @@ static const struct ov7670_reg ov7670_init_regtbl[] = {
 	{0xb8, 0x0a},
 };
 
+static int ov7670_update_reg(const struct device *dev, uint8_t reg, uint32_t mask, uint32_t new_val)
```
i2c_reg_update_byte_dt ? :)

iabdalkader replied:
Updated. I also fixed TSLB[3]. I tested RGB and Grayscale from YUV with these fixes, and confirm things work fine.

kartben replied:
Sorry if I wasn't clear, I meant you can just drop this function and use the built-in one from the I2C driver instead?

iabdalkader replied:
Wow I didn't know there's one. Okay.

kartben replied:
Yeah should have spotted the pattern earlier when I first suggested you factor some of the code, sorry. That's now a lot simpler than originally eh? :)

iabdalkader replied:
No worries, yes, it's much simpler now.

iabdalkader committed:
drivers: video: ov7670: Use default YUYV/RGB565 order.

TSLB[3] swaps the YUYV/RGB565 output.

Signed-off-by: Ibrahim Abdalkader <i.abdalkader@gmail.com>
